### PR TITLE
Fix shellcheck warnings/errors in cluster/gce/upgrade.sh

### PR DESCRIPTION
**Pull request by gavinfish**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Fix shellcheck failures in cluster/gce/upgrade.sh.

**Which issue(s) this PR fixes**:

Ref #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
none
```

### CODE REVIEW

File: cluster/gce/upgrade.sh [hunk 1]
```diff
@@ -241,24 +241,31 @@ function prepare-node-upgrade() {
   echo "== Preparing node upgrade (to ${KUBE_VERSION}). ==" >&2
   setup-base-image
 
-  SANITIZED_VERSION=$(echo ${KUBE_VERSION} | sed 's/[\.\+]/-/g')
+  SANITIZED_VERSION="${KUBE_VERSION}//[\.\+]/-/g"
```
Comment (php-coder) [inline cluster/gce/upgrade.sh]: This seems invalid expression. Perhaps, you mean something like this:
```suggestion
  SANITIZED_VERSION="${KUBE_VERSION//[\.\+]/-}"
```
?

File: cluster/gce/upgrade.sh [hunk 2]
```diff
@@ -241,24 +241,31 @@ function prepare-node-upgrade() {
...
+  export KUBE_PROXY_TOKEN
   NODE_PROBLEM_DETECTOR_TOKEN=$(get-env-val "${node_env}" "NODE_PROBLEM_DETECTOR_TOKEN")
   CA_CERT_BASE64=$(get-env-val "${node_env}" "CA_CERT")
+  export CA_CERT_BASE64
   EXTRA_DOCKER_OPTS=$(get-env-val "${node_env}" "EXTRA_DOCKER_OPTS")
+  export EXTRA_DOCKER_OPTS
   KUBELET_CERT_BASE64=$(get-env-val "${node_env}" "KUBELET_CERT")
+  export KUBELET_CERT_BASE64
   KUBELET_KEY_BASE64=$(get-env-val "${node_env}" "KUBELET_KEY")
+  export KUBELET_KEY_BASE64
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:268]: why are these exported? what needs them?

File: cluster/gce/upgrade.sh [hunk 3]
```diff
@@ -374,14 +387,15 @@ function do-node-upgrade() {
...
   #                 test/e2e/cluster_upgrade.go to match this EXACTLY.
-  local template_name=$(get-template-name-from-version ${SANITIZED_VERSION} ${NODE_INSTANCE_PREFIX})
+  local template_name
+  template_name=$(get-template-name-from-version "${SANITIZED_VERSION}" "${NODE_INSTANCE_PREFIX}")
   local old_templates=()
-  local updates=()
-  for group in ${INSTANCE_GROUPS[@]}; do
-    old_templates+=($(gcloud compute instance-groups managed list \
+  for group in "${INSTANCE_GROUPS[@]}"; do
+    old_templates=()
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:394]: I don't think this should be here, it's already initialized before the for loop, now you're resetting it each iteration of the loop

File: cluster/gce/upgrade.sh [hunk 4]
```diff
@@ -374,14 +387,15 @@ function do-node-upgrade() {
...
   # NOTE(zmerlynn): If you are changing this gcloud command, update
   #                 test/e2e/cluster_upgrade.go to match this EXACTLY.
-  local template_name=$(get-template-name-from-version ${SANITIZED_VERSION} ${NODE_INSTANCE_PREFIX})
+  local template_name
+  template_name=$(get-template-name-from-version "${SANITIZED_VERSION}" "${NODE_INSTANCE_PREFIX}")
   local old_templates=()
-  local updates=()
-  for group in ${INSTANCE_GROUPS[@]}; do
-    old_templates+=($(gcloud compute instance-groups managed list \
+  for group in "${INSTANCE_GROUPS[@]}"; do
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:393]: are you _sure_ this works exactly the same as before? I think we actually want globbing/splitting here

File: cluster/gce/upgrade.sh [hunk 5]
```diff
@@ -392,20 +406,20 @@ function do-node-upgrade() {
...
-      echo "${instances}"
+      echo "${instances[@]}"
       return ${list_instances_rc}
     fi
 
     process_count_left=${node_upgrade_parallelism}
     pids=()
     ret_code_sum=0  # Should stay 0 in the loop iff all parallel node upgrades succeed.
-    for instance in ${instances[@]}; do
+    for instance in "${instances[@]}"; do
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:422]: are you _sure_ this works exactly the same as before? I think we actually want globbing/splitting here

File: cluster/gce/upgrade.sh [hunk 6]
```diff
@@ -430,7 +444,7 @@ function do-node-upgrade() {
 
   # Remove the old templates.
   echo "== Deleting old templates in ${PROJECT}. ==" >&2
-  for tmpl in ${old_templates[@]}; do
+  for tmpl in "${old_templates[@]}"; do
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:447]: are you _sure_ this works exactly the same as before? I think we actually want globbing/splitting here

File: cluster/gce/upgrade.sh [hunk 7]
```diff
@@ -603,7 +617,7 @@ if [[ -z "${STORAGE_MEDIA_TYPE:-}" ]] && [[ "${STORAGE_BACKEND:-}" != "etcd2" ]]
   echo "export STORAGE_MEDIA_TYPE=application/json"
   echo ""
   if [ -t 0 ] && [ -t 1 ]; then
-    read -p "Would you like to continue with the new default, and lose the ability to downgrade to etcd2? [y/N] " confirm
+    read -r -p "Would you like to continue with the new default, and lose the ability to downgrade to etcd2? [y/N] " confirm
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:620]: why is the `-r` needed here?

File: cluster/gce/upgrade.sh [hunk 8]
```diff
@@ -638,7 +652,7 @@ if [[ "${master_upgrade}" == "true" ]]; then
     echo "In all cases, it is strongly recommended to have an etcd backup before upgrading."
     echo
     if [ -t 0 ] && [ -t 1 ]; then
-      read -p "Continue with default etcd version, which might upgrade etcd? [y/N] " confirm
+      read -r -p "Continue with default etcd version, which might upgrade etcd? [y/N] " confirm
```
Comment (spiffxp) [inline cluster/gce/upgrade.sh:655]: why is the `-r` needed here?

**Review by spiffxp — CHANGES_REQUESTED** [review]
changes to scripts inside of cluster/gce are risky, you need to help improve my confidence that these won't break things

File: cluster/gce/upgrade.sh (hunk 2 above)
Comment (gavinfish) [reply cluster/gce/upgrade.sh:268]: According to https://github.com/kubernetes/kubernetes/pull/7938#discussion_r29965211, these variables are required by code around https://github.com/kubernetes/kubernetes/blob/0926c9c47677e944c2dadcd7b39587866331eeba/cluster/gce/util.sh#L1182

File: cluster/gce/upgrade.sh (hunk 3 above)
Comment (gavinfish) [reply cluster/gce/upgrade.sh:394]: done

File: cluster/gce/upgrade.sh (hunk 7 above)
Comment (gavinfish) [reply cluster/gce/upgrade.sh:620]: refer to https://github.com/koalaman/shellcheck/wiki/SC2162

File: cluster/gce/upgrade.sh (hunk 4 above)
Comment (gavinfish) [reply cluster/gce/upgrade.sh:393]: No, they are not strictly equal. But I do not think globbing of individual elements is necessary here. refer to https://github.com/koalaman/shellcheck/wiki/SC2068

**Comment by gavinfish** [comment]
@spiffxp PTAL

/test pull-kubernetes-node-e2e

**Comment by spiffxp** [comment]
/approve
/lgtm
let's find out...

**Review by spiffxp — APPROVED** [review]

**Pull request merged**